apps/ledger-live-desktop/src/renderer/modals/UpdateFirmwareModal/index.tsx

```diff
@@ -106,17 +107,31 @@ const UpdateModal = ({
   const withFinal = useMemo(() => hasFinalFirmware(firmware?.final), [firmware]);
   const [cancel, setCancel] = useState<boolean>(false);
 
+  const isDisconnectedDeviceError = err instanceof DisconnectedDevice;
+  const isDisconnectedDeviceDuringOperationError = err instanceof DisconnectedDeviceDuringOperation;
+  const isDeviceDisconnected =
+    isDisconnectedDeviceError || isDisconnectedDeviceDuringOperationError;
+
+  console.log("DEVICE", deviceModelId, props.device, updatedDeviceInfo);
+
   const onRequestCancel = useCallback(() => {
-    showDisclaimer || stateStepId === "finish" ? onRequestClose() : setCancel(state => !state);
-  }, [showDisclaimer, stateStepId, onRequestClose]);
+    (showDisclaimer && !isDeviceDisconnected) || stateStepId === "finish" || cancel
+      ? onRequestClose()
+      : setCancel(true);
+  }, [cancel, showDisclaimer, isDeviceDisconnected, stateStepId, onRequestClose]);
 
   const createSteps = useCallback(
     ({ withResetStep }: { withResetStep: boolean }) => {
       const hasRestoreStep =
         firmware && isDeviceLocalizationSupported(firmware.final.name, deviceModelId);
-      const restoreStepLabel = t("manager.modal.steps.restore");
-      const installUpdateLabel =
+
+      const restoreStepLabel =
```

Comment:
I know it was already done like this, but I think we should extract this logic outside of this component with a function that builds steps.

Reply:
We don't need a translation function at this stage, we could do it later in the component rendering:
label={t(step.label) as string}